tiledb/sm/serialization/query.cc

```diff
@@ -261,84 +261,122 @@ Status subarray_to_capnp(
   return Status::Ok();
 }
 
-Status subarray_from_capnp(
-    const capnp::Subarray::Reader& reader, Subarray* subarray) {
-  RETURN_NOT_OK(subarray->set_coalesce_ranges(reader.getCoalesceRanges()));
+Subarray subarray_from_capnp(
+    const capnp::Subarray::Reader& reader,
+    const Array* array,
+    Layout layout,
+    stats::Stats* parent_stats,
+    shared_ptr<Logger> logger) {
+  bool coalesce_ranges = reader.getCoalesceRanges();
   auto ranges_reader = reader.getRanges();
+
   uint32_t dim_num = ranges_reader.size();
+  std::vector<RangeSetAndSuperset> range_subset;
+  range_subset.reserve(dim_num);
+  std::vector<bool> is_default(dim_num, false);
   for (uint32_t i = 0; i < dim_num; i++) {
     auto range_reader = ranges_reader[i];
-
-    auto data = range_reader.getBuffer();
-    auto data_ptr = data.asBytes();
-    if (range_reader.hasBufferSizes()) {
-      auto ranges = range_buffers_from_capnp(range_reader);
-      RETURN_NOT_OK(subarray->set_ranges_for_dim(i, ranges));
-
-      // Set default indicator
-      subarray->set_is_default(i, range_reader.getHasDefaultRange());
+    Datatype type = datatype_enum(range_reader.getType());
```

Comment:
This line should be calling something like `subarray_ranges_from_capnp`. It would be passed `range_reader` and return an object suitable for initializing with the new C.41 construction, which in this case should be `RangeSetAndSuperset`. While this isn't strictly necessary, it would make the code easier to read by delineating dependent lists etc. more clearly.